plugins/inputs/prometheus/consul.go

```diff
@@ -0,0 +1,174 @@
+package prometheus
+
+import (
+	"bytes"
+	"context"
+	"log"
+	"net/url"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+)
+
+type ConsulServiceInfo struct {
+	ID             string
+	Node           string
+	Address        string
+	Datacenter     string
+	NodeMeta       map[string]string
+	ServiceID      string
+	ServiceName    string
+	ServiceAddress string
+	ServiceTags    []string
+	ServiceMeta    map[string]string
+	ServicePort    int
+}
+
+func convertConsulService(s *api.CatalogService) *ConsulServiceInfo {
+	return &ConsulServiceInfo{
+		ID:             s.ID,
+		Node:           s.Node,
+		Address:        s.Address,
+		Datacenter:     s.Datacenter,
+		NodeMeta:       s.NodeMeta,
+		ServiceID:      s.ServiceID,
+		ServiceAddress: s.ServiceAddress,
+		ServiceTags:    s.ServiceTags,
+		ServiceMeta:    s.ServiceMeta,
+		ServicePort:    s.ServicePort,
+	}
+}
+
+func (p *Prometheus) startConsul(ctx context.Context) error {
+	config := api.DefaultConfig()
+	if p.ConsulConfig.Agent != "" {
+		config.Address = p.ConsulConfig.Agent
+	}
+
+	consul, err := api.NewClient(config)
+	if err != nil {
+		log.Printf("E! [inputs.prometheus] Cannot connect to the Consul agent. Error: %s", err)
+		return nil
+	}
+
+	// Parse the template for metrics URL
+	for i, q := range p.ConsulConfig.Queries {
+		serviceURLTemplate, err := template.New("URL").Parse(q.ServiceURL)
+		if err != nil {
+			log.Printf("E! [inputs.prometheus] Could not parse the Consul query URL template (%s), skipping it. Error: %s", q.ServiceURL, err)
+			continue
+		}
+		p.ConsulConfig.Queries[i].serviceURLTemplate = serviceURLTemplate
+
+		// Allow to use join function in tags
+		templateFunctions := template.FuncMap{"join": strings.Join}
+		// Parse the tag value templates
+		p.ConsulConfig.Queries[i].serviceExtraTagsTemplate = make(map[string]*template.Template)
+		for tagName, tagTemplateString := range q.ServiceExtraTags {
+			tagTemplate, err := template.New(tagName).Funcs(templateFunctions).Parse(tagTemplateString)
+			if err != nil {
+				log.Printf("E! [inputs.prometheus] Could not parse the Consul query Extra Tag template (%s), skipping it. Error: %s", tagTemplateString, err)
+				continue
+			}
+
+			p.ConsulConfig.Queries[i].serviceExtraTagsTemplate[tagName] = tagTemplate
+		}
+	}
+
+	catalog := consul.Catalog()
+
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+		err := p.refreshConsulServices(catalog)
+		if err != nil {
+			log.Printf("E! [inputs.prometheus] Unable to refresh Consul services: %v", err)
+		}
```

Comment:
Does it really make sense to unconditionally continue here?

Reply:
Consul might be available later and services will start to be queried without the need to restart telegraf.